// webbox/Server.js
/**
 * Created by Артём on 14.10.2016.
 */
var colors, express, fs, pug, pathmodule;
var modules = ["colors/safe", 'express', 'fs', 'pug', 'path'];

function Box(install_modules) {
    colors = require('colors/safe');    // Я не уверен нужны ли здесь эти 3 модуля,
    express = require('express');       //т.к. ты все равно их записываешь в
    fs = require('fs');
    pug = require('pug');               //peerDependencies
    pathmodule = require('path');

    console.log(colors.cyan.bold('[WP] Starting WebBox.'));

    this.app = express(); 	// Include Express
    this.port = 80; 		// Default listening port
    this.path = pathmodule.dirname(require.main.filename); // Get project path
    this.module_path_abs = __dirname; // Get module path
    this.module_path = "/node_modules/webbox/";
}

Box.prototype.createServer = function(srv_port) {
    this.port = srv_port;

    // Start listening port
    this.app.listen(this.port);
    console.log(colors.green('[WP] Server listening on port *:' + this.port));

    // Create API page
    this.newResourceFolder(this.module_path+'css/');
    this.newResourceFolder(this.module_path+'imgs/');
    this.newResourceFolder(this.module_path+'downloads/');
    this.newStaticPage('/cat-api', this.module_path+'api.html');
};

Box.prototype.newResourceFolder = function (dir) {
    var abs_path = this.path + '/';
    this.app.get(dir+':name', function (req, res, next) {
        var pth = abs_path + dir + req.params.name;
        res.sendFile(pth);
    });
    console.log('[WP] New resource folder: ' + dir);
};

Box.prototype.newDownloadFolder = function (dir) {
    var abs_path = this.path + '/';
    this.app.get(dir+':name', function (req, res, next) {
        var pth = abs_path + dir + req.params.name;
        res.sendFile(pth);
    });
    console.log('[WP] New download folder: ' + dir);
}

/*
 Creates a new request listener with user-function.
 */
Box.prototype.newRequestListener = function(req_path, method, callback) {
    if (method.toLowerCase() == "get") {
        this.app.get(req_path, function(req,res) {
            callback(req, res);
        });
    }
    else if (method.toLowerCase() == "post") {
        this.app.get(req_path, function(req,res) {
            callback(req, res);
        });
    }
}

/*
 Creates a new request listener with static page.
 */
Platform.prototype.newStaticPage = function (req_path, page_path) {
    var abs_path = this.path + '/';

    this.app.get(req_path, function (req, res, next) {
        var pth = abs_path + page_path;
        res.sendFile(pth);
    });
    console.log(colors.yellow('[WP] New static page: ' + page_path));
}



Box.prototype.redirectLink = function(res, path)
{
    res.redirect(path);
}

exports.Box=Box;
